### PR TITLE
Partial support for reading CXSMILES

### PR DESCRIPTION
**Pull request by greglandrum**
Beginning of the work for #1226. This isn't done, but it's probably already useful.

This PR includes:
 - molecule name
 - atom labels
 - coordinate bonds
 - query atoms
 - coordinates
 - radicals

Additional stuff in here:
 - a new version of SmilesToMol() that takes a parameters object.
 - removal of a bunch of compiler warnings on windows (I was doing the work there and couldn't help myself)

### CODE REVIEW

**Comment by bp-kelley** [comment]
Hah, neat.  This will take some time to review.

**Comment by bp-kelley** [comment]
@greglandrum Im' trying to suss out a good review scheme for this that isn't just replicating your test suite.  This might take a bit of time, or we can accept the PR and just build on it.  I'm fine either way.

**Comment by greglandrum** [comment]
If you think the initial feature coverage isn't useless and the code doesn't look too bad, then let's start here and then figure out what should come next?

**Comment by greglandrum** [comment]
@bp-kelley : any objection to this being merged?

**Comment by bp-kelley** [comment]
Give me up the weekend to feel a bit more comfortable with it.  It won't stop me from merging btw :)

**Comment by greglandrum** [comment]
Sounds good, thanks

File: Code/GraphMol/QueryOps.cpp [hunk 1]
```diff
@@ -331,20 +330,130 @@ ATOM_EQUALS_QUERY *makeAtomInRingQuery() {
...
+}
+ATOM_OR_QUERY *makeMHAtomQuery() {
+  // using the definition from Marvin Sketch, which produces the following SMARTS:
+  // !#2!#5!#6!#7!#8!#9!#10!#14!#15!#16!#17!#18!#33!#34!#35!#36!#52!#53!#54!#85!#86
+  // it's easier to define what isn't a metal than what is. :-)
+  ATOM_OR_QUERY *res = new ATOM_OR_QUERY;
+  res->setDescription("AtomOr");
+  res->setNegation(true);
+  res->addChild(
+    Queries::Query<int, Atom const *, true>::CHILD_TYPE(makeAtomNumQuery(2)));
```
Comment (bp-kelley) [inline Code/GraphMol/QueryOps.cpp]: Sure looks like you could use a makeAtomInRange query here :)

File: Code/GraphMol/SmilesParse/CXSmilesOps.cpp [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  ++first;
+  unsigned int atIdx = 0;
+  while (first != last && *first != ')') {
+    RDGeom::Point3D pt;
+    std::string tkn = read_text_to(first, last, ';', ')');
+    if (tkn != "") {
+      std::vector<std::string> tokens;
+      boost::split(tokens, tkn, boost::is_any_of(std::string(",")));
+      if (tokens.size() >= 1 && tokens[0].size())
+        pt.x = boost::lexical_cast<double>(tokens[0]);
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/CXSmilesOps.cpp]: I suppose I would like some exception catching here for parse errors.  I feel like I should make some explicitly broken CX smiles for testing

File: Code/GraphMol/SmilesParse/CXSmilesOps.cpp [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  }
+  return true;
+}
+
+template <typename Iterator>
+bool parse_it(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  if (first >= last || *first != '|') return false;
+  ++first;
+  while (first < last && *first != '|') {
+    if (*first == '(') {
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/CXSmilesOps.cpp]: good place for a switch statement.

File: Code/GraphMol/SmilesParse/SmartsWrite.cpp [hunk 4]
```diff
@@ -30,12 +30,12 @@ std::string _recurseBondSmarts(const Bond *bond,
...
 
 bool _checkForOrAndLowAnd(std::string smarts) {
-  int orLoc, andLoc;
+  size_t orLoc, andLoc;
   // if we're a pure recursive smarts, we don't need to worry about this
   if (smarts[0] == '$' && smarts[smarts.size() - 1] == ')') return false;
   orLoc = smarts.find(",");
   andLoc = smarts.find(";");
-  if ((orLoc > 0) && (andLoc > 0)) {
+  if ((orLoc != std::string::npos) && (andLoc != std::string::npos)) {
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/SmartsWrite.cpp]: good catch.

File: Code/GraphMol/SmilesParse/SmilesParse.cpp [hunk 5]
```diff
@@ -200,23 +217,38 @@ RWMol *SmilesToMol(const std::string &smiles, int debugParse, bool sanitize,
...
   }
-
+  if (res && params.allowCXSMILES && cxPart != "") {
+    // it's goofy that we have to do this, but c++0x doesn't seem to have
+    // a way to get a const_iterator from a non-const std::string. In C++11
+    // we could just use .cend()
+    const std::string &cxcopy = cxPart;
+    std::string::const_iterator pos;
+    SmilesParseOps::parseCXExtensions(*res, cxcopy, pos);
+    if (params.parseName && pos != cxcopy.end()) {
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/SmilesParse.cpp]: Yay!

File: Code/GraphMol/SmilesParse/cxsmiles_test.cpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+    TEST_ASSERT(m->getNumAtoms() == 3);
+    TEST_ASSERT(m->getAtomWithIdx(0)->getProp<std::string>(
+                    common_properties::atomLabel) == "foo");
+    TEST_ASSERT(m->getAtomWithIdx(2)->getProp<std::string>(
+                    common_properties::atomLabel) == "bar");
+    TEST_ASSERT(!m->getAtomWithIdx(1)->hasProp(common_properties::atomLabel));
+    delete m;
+  }
+  {  // attachment points, example from the docs
+    std::string smiles = "C[C@H](N*)C(*)=O |$;;;_AP1;;_AP2;$|";
```
Comment (bp-kelley) [inline Code/GraphMol/SmilesParse/cxsmiles_test.cpp]: cx smiles are ugly.

**Comment by bp-kelley** [comment]
Ok.  Looks good.  I'll merge tomorrow morning after you see the comments.

One comment is that we need to parse some bad cx smiles to test failure modes.

**Comment by greglandrum** [comment]
Agreed about having some bad ones. Particularly real-world bad examples.

**Pull request merged**